server/src/main/java/org/elasticsearch/index/mapper/KeywordFieldMapper.java

```diff
@@ -417,65 +428,40 @@ public KeywordFieldMapper build(MapperBuilderContext context) {
+            if (useDocValuesSparseIndex
                 && indexCreatedVersion.onOrAfter(IndexVersions.HOSTNAME_DOC_VALUES_SPARSE_INDEX)
-                && shouldUseDocValuesSparseIndex(indexSortConfig, indexMode, fullFieldName)) {
+                && shouldUseDocValuesSparseIndex(hasDocValues.getValue(), indexSortConfig, indexMode, fullFieldName)) {
                 return new FieldType(Defaults.FIELD_TYPE_WITH_SKIP_DOC_VALUES);
             }
             return new FieldType(Defaults.FIELD_TYPE);
         }
 
-        /**
-         * Determines whether to use a sparse index representation for doc values.
-         *
-         * <p>If the field is explicitly indexed by setting {@code index: true}, we do not use
-         * a sparse doc values index but instead rely on the inverted index, as is typically
-         * the case for keyword fields.</p>
-         *
-         * <p>This method checks several conditions to decide if the sparse index format
-         * should be applied:</p>
-         *
-         * <ul>
-         *     <li>Returns {@code false} immediately if the field is explicitly indexed.</li>
-         *     <li>Ensures the field is not explicitly configured as indexed (i.e., {@code index} has its default value).</li>
-         *     <li>Requires doc values to be enabled.</li>
-         *     <li>Index mode must be {@link IndexMode#LOGSDB}.</li>
-         *     <li>Field name must be {@code host.name}.</li>
-         *     <li>The {@code host.name} field must be a primary sort field.</li>
-         * </ul>
-         *
-         * <p>Returns {@code true} if all conditions are met, indicating that sparse doc values
-         * should be used. Otherwise, returns {@code false}.</p>
-         *
-         * @param indexSortConfig The index sort configuration, used to check primary sorting.
-         * @param indexMode The mode of the index, which must be {@link IndexMode#LOGSDB}.
-         * @param fullFieldName The name of the field being checked, which must be {@code host.name}.
-         * @return {@code true} if sparse doc values should be used, otherwise {@code false}.
-         */
-
-        private boolean shouldUseDocValuesSparseIndex(
+        private static boolean shouldUseDocValuesSparseIndex(
+            final boolean hasDocValues,
             final IndexSortConfig indexSortConfig,
             final IndexMode indexMode,
             final String fullFieldName
         ) {
-            if (indexed.isSet() && indexed.getValue()) {
-                return false;
-            }
-            return indexed.isConfigured() == false
-                && hasDocValues.getValue()
+            return hasDocValues
                 && IndexMode.LOGSDB.equals(indexMode)
                 && HOST_NAME.equals(fullFieldName)
-                && (indexSortConfig != null && indexSortConfig.hasPrimarySortOnField(HOST_NAME));
+                && indexSortConfigByHostName(indexSortConfig);
```

Comment:
Here I assume we would like to create the sparse index on `host.name` no matter if it is the first field or not.